Handle stripe auth failure better

Seen in prod on a key auth failure. This should probably be handled cleaner and dump less unnecessary data into the logs.

    [ERROR] dshop.routes.stripe: Failed to verify stripe credentials { Error: Invalid API Key provided: [censored]
        at IncomingMessage.res.once (/app/node_modules/stripe/lib/StripeResource.js:184:21)
        at Object.onceWrapper (events.js:286:20)
        at IncomingMessage.emit (events.js:203:15)
        at IncomingMessage.EventEmitter.emit (domain.js:466:23)
        at endReadableNT (_stream_readable.js:1145:12)
        at process._tickCallback (internal/process/next_tick.js:63:19)
      type: 'StripeAuthenticationError',
      raw:
       { message: 'Invalid API Key provided: [censored]',
         type: 'invalid_request_error',
         headers:
          { server: 'nginx',
            date: 'Wed, 29 Jul 2020 17:53:40 GMT',
            'content-type': 'application/json',
            'content-length': '109',
            connection: 'keep-alive',
            'access-control-allow-credentials': 'true',
            'access-control-allow-methods': 'GET, POST, HEAD, OPTIONS, DELETE',
            'access-control-allow-origin': '*',
            'access-control-expose-headers':
             'Request-Id, Stripe-Manage-Version, X-Stripe-External-Auth-Required, X-Stripe-Privileged-Session-Required',
            'access-control-max-age': '300',
            'cache-control': 'no-cache, no-store',
            'www-authenticate': 'Bearer realm="Stripe"',
            'strict-transport-security': 'max-age=31556926; includeSubDomains; preload' },
         statusCode: 401,
         requestId: undefined },
      rawType: 'invalid_request_error',
      code: undefined,
      doc_url: undefined,
      param: undefined,
      detail: undefined,
      headers:
       { server: 'nginx',
         date: 'Wed, 29 Jul 2020 17:53:40 GMT',
         'content-type': 'application/json',
         'content-length': '109',
         connection: 'keep-alive',
         'access-control-allow-credentials': 'true',
         'access-control-allow-methods': 'GET, POST, HEAD, OPTIONS, DELETE',
         'access-control-allow-origin': '*',
         'access-control-expose-headers':
          'Request-Id, Stripe-Manage-Version, X-Stripe-External-Auth-Required, X-Stripe-Privileged-Session-Required',
         'access-control-max-age': '300',
         'cache-control': 'no-cache, no-store',
         'www-authenticate': 'Bearer realm="Stripe"',
         'strict-transport-security': 'max-age=31556926; includeSubDomains; preload' },
      requestId: undefined,
      statusCode: 401,
      charge: undefined,
      decline_code: undefined,
      payment_intent: undefined,
      payment_method: undefined,
      setup_intent: undefined,
      source: undefined }

Rel: #325

--- a/backend/routes/stripe.js
+++ b/backend/routes/stripe.js
@@ -4,10 +4,7 @@ const bodyParser = require('body-parser')
 const randomstring = require('randomstring')
 const Stripe = require('stripe')
 
-const {
-  validatePublishableKey,
-  validateSecretKey
-} = require('@origin/utils/stripe')
+const { validateStripeKeys } = require('@origin/utils/stripe')
 
 const { Shop, ExternalPayment, Network } = require('../models')
 const { authShop } = require('./_auth')
@@ -201,7 +198,12 @@ module.exports = function (router) {
     let valid = false
     const { stripeKey, stripeBackend } = req.body
 
-    if (validatePublishableKey(stripeKey) && validateSecretKey(stripeBackend)) {
+    if (
+      validateStripeKeys({
+        publishableKey: stripeKey,
+        secretKey: stripeBackend
+      })
+    ) {
       try {
         const stripe = Stripe(stripeBackend)
 
@@ -211,7 +213,6 @@ module.exports = function (router) {
       } catch (err) {
         log.error('Failed to verify stripe credentials')
         log.debug(err)
-        valid = false
       }
     }
 

--- a/backend/routes/stripe.js
+++ b/backend/routes/stripe.js
@@ -4,6 +4,11 @@ const bodyParser = require('body-parser')
 const randomstring = require('randomstring')
 const Stripe = require('stripe')
 
+const {
+  validatePublishableKey,
+  validateSecretKey
+} = require('@origin/utils/stripe')
+
 const { Shop, ExternalPayment, Network } = require('../models')
 const { authShop } = require('./_auth')
 const { getConfig } = require('../utils/encryptedConfig')
@@ -194,18 +199,20 @@ module.exports = function (router) {
 
   router.post('/stripe/check-creds', authShop, async (req, res) => {
     let valid = false
+    const { stripeKey, stripeBackend } = req.body
 
-    try {
-      const { stripeBackend } = req.body
+    if (validatePublishableKey(stripeKey) && validateSecretKey(stripeBackend)) {
+      try {
+        const stripe = Stripe(stripeBackend)
 
-      const stripe = Stripe(stripeBackend)
+        await stripe.customers.list({ limit: 1 })
 
-      await stripe.customers.list({ limit: 1 })
-
-      valid = true
-    } catch (err) {
-      log.error('Failed to verify stripe credentials', err)
-      valid = false
+        valid = true
+      } catch (err) {
+        log.error('Failed to verify stripe credentials')
+        log.debug(err)
+        valid = false
+      }
     }
 
     return res.status(200).send({

--- a/backend/utils/stripe.js
+++ b/backend/utils/stripe.js
@@ -1,25 +1,5 @@
-const { getLogger } = require('./logger')
+const { normalizeDescriptor } = require('@origin/utils/stripe')
 
-const log = getLogger('utils.stripe')
-
-const DEFAULT_DESCRIPTOR = 'DSHOP'
-const DESCRIPTOR_INVALID_CHARS = /[^A-Z ]/gi
-
-/**
- * Normalize a "statement descriptor" for stripe payments.
- *
- * @see {@link https://stripe.com/docs/statement-descriptors}
- * @param d {string} - to normalie
- * @returns {string} normalized
- */
-function normalizeDescriptor(d) {
-  // Go to default if it's something we can't do anything with
-  if (!d || typeof d !== 'string' || d.length < 5) {
-    log.warn('Statement descriptor is missing or invalid')
-    return DEFAULT_DESCRIPTOR
-  }
-
-  return d.replace(DESCRIPTOR_INVALID_CHARS, '')
-}
+// TODO: remove this compat file
 
 module.exports = { normalizeDescriptor }

--- a/packages/utils/stripe.js
+++ b/packages/utils/stripe.js
@@ -39,8 +39,30 @@ function validateSecretKey(key) {
   return SECRET_KEY_PREFIXES.some((prefix) => key.startsWith(prefix))
 }
 
+/**
+ * Validate secret key
+ *
+ * @param args {object} - args
+ * @param args.publishableKey {string} - The Strie publishable key
+ * @param args.secretKey {string} - The Strie secret key
+ * @returns {boolean} - if the keys appear to be valid
+ */
+function validateStripeKeys({ publishableKey, secretKey }) {
+  if (!publishableKey || !secretKey) {
+    return false
+  }
+
+  // Make sure they're not a mix of live and test keys
+  if (publishableKey.includes('live') && !secretKey.includes('live')) {
+    return false
+  }
+
+  return validatePublishableKey(publishableKey) && validateSecretKey(secretKey)
+}
+
 module.exports = {
   normalizeDescriptor,
   validatePublishableKey,
-  validateSecretKey
+  validateSecretKey,
+  validateStripeKeys
 }

--- a/packages/utils/stripe.js
+++ b/packages/utils/stripe.js
@@ -1,0 +1,46 @@
+const DEFAULT_DESCRIPTOR = 'DSHOP'
+const DESCRIPTOR_INVALID_CHARS = /[^A-Z ]/gi
+const PUBLIC_KEY_PREFIXES = ['pk_test_', 'pk_live_']
+const SECRET_KEY_PREFIXES = ['sk_test_', 'sk_live_']
+
+/**
+ * Normalize a "statement descriptor" for stripe payments.
+ *
+ * @see {@link https://stripe.com/docs/statement-descriptors}
+ * @param d {string} - to normalie
+ * @returns {string} normalized
+ */
+function normalizeDescriptor(d) {
+  // Go to default if it's something we can't do anything with
+  if (!d || typeof d !== 'string' || d.length < 5) {
+    return DEFAULT_DESCRIPTOR
+  }
+
+  return d.replace(DESCRIPTOR_INVALID_CHARS, '')
+}
+
+/**
+ * Validate publishable key
+ *
+ * @param key {string} - the key to validate
+ * @returns {boolean} - if the key appears valid
+ */
+function validatePublishableKey(key) {
+  return PUBLIC_KEY_PREFIXES.some((prefix) => key.startsWith(prefix))
+}
+
+/**
+ * Validate secret key
+ *
+ * @param key {string} - the key to validate
+ * @returns {boolean} - if the key appears valid
+ */
+function validateSecretKey(key) {
+  return SECRET_KEY_PREFIXES.some((prefix) => key.startsWith(prefix))
+}
+
+module.exports = {
+  normalizeDescriptor,
+  validatePublishableKey,
+  validateSecretKey
+}

--- a/shop/src/pages/admin/settings/payments/StripeModal.js
+++ b/shop/src/pages/admin/settings/payments/StripeModal.js
@@ -2,6 +2,7 @@ import React, { useReducer } from 'react'
 
 import pick from 'lodash/pick'
 import pickBy from 'lodash/pickBy'
+import { validatePublishableKey, validateSecretKey } from '@origin/utils/stripe'
 
 import { formInput, formFeedback } from 'utils/formHelpers'
 import ConnectModal from './_ConnectModal'
@@ -23,11 +24,11 @@ const initialState = {
 const validate = (state) => {
   const newState = {}
 
-  if (!state.stripeBackend) {
+  if (!state.stripeBackend || !validateSecretKey(state.stripeBackend)) {
     newState.stripeBackendError = 'Secret key is required'
   }
 
-  if (!state.stripeKey) {
+  if (!state.stripeKey || !validatePublishableKey(state.stripeKey)) {
     newState.stripeKeyError = 'Client key is required'
   }
 

--- a/shop/src/pages/admin/settings/payments/StripeModal.js
+++ b/shop/src/pages/admin/settings/payments/StripeModal.js
@@ -2,7 +2,11 @@ import React, { useReducer } from 'react'
 
 import pick from 'lodash/pick'
 import pickBy from 'lodash/pickBy'
-import { validatePublishableKey, validateSecretKey } from '@origin/utils/stripe'
+import {
+  validateStripeKeys,
+  validatePublishableKey,
+  validateSecretKey
+} from '@origin/utils/stripe'
 
 import { formInput, formFeedback } from 'utils/formHelpers'
 import ConnectModal from './_ConnectModal'
@@ -30,6 +34,19 @@ const validate = (state) => {
 
   if (!state.stripeKey || !validatePublishableKey(state.stripeKey)) {
     newState.stripeKeyError = 'Client key is required'
+  }
+
+  // Validate them as a pair
+  if (
+    !newState.stripeKeyError &&
+    !newState.stripeBackendError &&
+    !validateStripeKeys({
+      publishableKey: state.stripeKey,
+      secretKey: state.stripeBackend
+    })
+  ) {
+    newState.stripeKeyError = "Stripe keys don't match"
+    newState.stripeBackendError = "Stripe keys don't match"
   }
 
   const valid = Object.keys(newState).every((f) => !f.endsWith('Error'))

